core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1252,6 +1254,28 @@ private[spark] class DAGScheduler(
     execCores.map(cores => properties.setProperty(EXECUTOR_CORES_LOCAL_PROPERTY, cores))
   }
 
+  /**
+   * If push based shuffle is enabled, set the shuffle services to be used for the given
+   * shuffle map stage. The list of shuffle services is determined based on the list of
+   * active executors tracked by block manager master at the start of the stage.
+   */
+  private def prepareShuffleServicesForShuffleMapStage(stage: ShuffleMapStage) {
```

Comment:
We also want to get community inputs on this.
This PR provides the implementation for shuffle service (merger) location selection logic for YARN with support for dynamic resource allocation.
The selected shuffle service locations will be provided to `ShuffleMapTask` via `ShuffleDependency`, so that all mappers can push blocks to corresponding shuffle services.
We think it would be better to make the `getMergerLocations` API pluggable, so that in addition to this default implementation for YARN with DRA support, people can provide their own implementations based on their unique cluster deployment setup.
For example, in cluster deployment with disaggregated shuffle services, where the shuffle service availability to a given Spark application is not as dynamic as the case of YARN with DRA, the pluggable API would enable a very different strategy for selecting merger locations.

Want to get inputs from the community for this pluggable merger selection API, and the parameters that would fit into such an API.
CC @Ngone51 @jiangxb1987 @attilapiros @tgravescs @mridulm

Reply:
How about a subinterface extending `ShuffleDriverComponents` for push based shuffle - with a suitable default implementation ?

Reply:
I definitely agree it should be pluggable. ShuffleDriverComponents seems like a decent place to put it, although that code is still under flux so not sure how it will go with mixing right now. Right now its using some pretty strict logic from the block manager master so I wonder how difficult writing your own will be. More or less are there enough public api's (BlockManagerMaster is not public)- or do you have to be in spark package namespace to write one.  Perhaps this should be a separate followup issue?

Reply:
I have created a sub-task to track this for now - https://issues.apache.org/jira/browse/SPARK-33329